File: minimal_python/MinimalApp.py
import sys
from AppClient import AppClient
from DataTypes.Matrix44 import Matrix44
from robotcontrolapp_pb2 import ButtonState, CheckboxState


class MinimalApp(AppClient):
    # Initializes the app. Pass the app name (as defined in rcapp.xml) and socket to connect to (default: "localhost:5000")
    def __init__(self, appName: str, target: str):
        AppClient.__init__(self, appName, target)

        # Gets called on remote app function calls received from the robot control

    def _AppFunctionHandler(self, function):
        # This prints the received function data
        print("Received app function:")
        print(function)

        # Select the app function to call
        if function.name == "pow":
            self.ExampleExponentiation(function)
        else:
            # Example on how to handle app functions (these are called by the program command "App")
            self.ExamplePrintAppFunctionParameters(function)

        # Confirm that the function finished, otherwise the robot program will wait forever.
        # You may send this later if the function call takes some time but it must be sent at some point.
        self.SendFunctionDone(function.call_id)

    # Gets called on remote UI update requests received from the robot control
    def _UiUpdateHandler(self, updates):
        # This prints the received UI updates:
        print("Received UI updates: ")
        print(updates)

        # Example on how to handle UI events (button clicked, value changed etc)
        self.ExamplePrintUIEvents(updates)

        # Example on how to handle UI events (button clicked, value changed etc)
        self.ExampleUIElementClicked(updates)

    # This example prints all app functions contained in an AppFunction request (e.g. coming from the App program command)
    def ExamplePrintAppFunctionParameters(self, function):
        print(
            f"App function '{function.name}' called with call ID {function.call_id}, label = '{function.label}', ui hint = '{function.ui_hint}', number of parameters = {len(function.parameters)}"
        )

        # print all parameters
        for parameter in function.parameters:
            if parameter.HasField("bool_value"):
                # boolean
                print(
                    f"\tparameter '{parameter.name}', type 'bool' {parameter.bool_value}"
                )
            elif parameter.HasField("int64_value"):
                # integer
                print(
                    f"\tparameter '{parameter.name}', type 'int64' {parameter.int64_value}"
                )
            elif parameter.HasField("double_value"):
                # double
                print(
                    f"\tparameter '{parameter.name}', type 'double' {parameter.double_value:.2f}"
                )
            elif parameter.HasField("string_value"):
                # string
                print(
                    f"\tparameter '{parameter.name}', type 'string' {parameter.string_value}"
                )
            elif parameter.HasField("vector3_value"):
                # 3 dimensional vector
                print(
                    f"\tparameter '{parameter.name}', type 'vector3' ({parameter.vector3_value.GetX():.2f} {parameter.vector3_value.GetY():.2f} {parameter.vector3_value.GetZ():.2f})"
                )
            elif parameter.HasField("cartesian_value"):
                # cartesian position and orientation
                matrix = Matrix44(parameter.cartesian_value)
                print(
                    f"\tparameter '{parameter.name}', type 'cartesian' X={matrix.GetX():.2f}, Y={matrix.GetY():.2f}, Z={matrix.GetZ():.2f}, A={matrix.GetA():.2f}, B={matrix.GetB():.2f}, C={matrix.GetC():.2f}"
                )

    # This example exponentiates the value of a number variable by a given exponent and assigns the result to a different number variable
    def ExampleExponentiation(self, function):
        # Get the function parameters by iterating over the list
        for parameter in function.parameters:
            if parameter.name == "base_variable" and parameter.HasField("string_value"):
                baseVariableName = parameter.string_value
            elif parameter.name == "result_variable" and parameter.HasField(
                "string_value"
            ):
                resultVariableName = parameter.string_value
            elif parameter.name == "exponent_number" and parameter.HasField(
                "double_value"
            ):
                exponentValue = parameter.double_value

        try:
            # Now request the value of the base number variable from the robot control
            baseVariable = self.GetNumberVariable(baseVariableName)
            baseValue = baseVariable.GetValue()

            # And calculate the result
            resultValue = pow(baseValue, exponentValue)

            # Write the result to the target variable
            self.SetNumber(resultVariableName, resultValue)

            print(
                f'Calculated {baseValue}^{exponentValue} = {resultValue}, result was written to variable "{resultVariableName}"'
            )
        except Exception as ex:
            print(f'Function call "exponentiation" failed', file=sys.stderr)
            print(ex, file=sys.stderr)

    # This example prints all UI events contained in a UI update
    def ExamplePrintUIEvents(self, updates):
        for update in updates:
            if update.state.HasField("button_state"):
                # button clicked
                isClicked = update.state.button_state == ButtonState.CLICKED
                print(
                    f"Button '{update.element_name}' changed: is clicked = {isClicked}"
                )
            elif update.state.HasField("checkbox_state"):
                # checkbox toggled
                isChecked = update.state.checkbox_state == CheckboxState.CHECKED
                print(
                    f"Checkbox '{update.element_name}' changed: is checked = {isChecked}"
                )
            elif update.state.HasField("dropdown_state"):
                # drop down selected
                selectedOption = update.state.dropdown_state.selected_option
                print(
                    f"Dropdown '{update.element_name}' changed: selected option = {selectedOption}"
                )
            elif update.state.HasField("textfield_state"):
                # text box changed
                text = update.state.textfield_state.current_text
                print(f"Text box '{update.element_name}' changed: text = {text}")
            elif update.state.HasField("numberfield_state"):
                # number box changed
                number = update.state.numberfield_state.current_number
                print(f"Number box '{update.element_name}' changed: value = {number}")
            elif update.state.HasField("image_state"):
                # image clicked
                if update.state.image_state.HasField("clicked_at"):
                    x = update.state.image_state.clicked_at.X()
                    y = update.state.image_state.clicked_at.X()
                    isClicked = update.state.image_state.clicked_at.is_clicked
                    if isClicked:
                        print(
                            f"Image clicked '{update.element_name}': x = {x:.3f}, y = {y:.3f}"
                        )

    # Counting variable for ExampleUIElementClicked()
    _examplePlusMinusValue = 0

    # This example handles the plus/minus buttons to increase a text element in the app UI
    def ExampleUIElementClicked(self, updates):
        for update in updates:
            # The following sample will increase/decrease a value if the plus/minus button is pressed and send it back to the UI.
            if update.element_name == "buttonMinus":  # button minus pressed
                self._examplePlusMinusValue -= 1
                print(f"Decreasing UI number to {self._examplePlusMinusValue:.0f}")
                self.SetNumber("textLeftRight", self._examplePlusMinusValue)
            elif update.element_name == "buttonPlus":  # button plus pressed
                self._examplePlusMinusValue += 1
                print(f"Increasing UI number to {self._examplePlusMinusValue:.0f}")
                self.SetNumber("textLeftRight", self._examplePlusMinusValue)

File: minimal_cpp/src/MinimalApp.cpp
#include "MinimalApp.h"

#include <math.h>
#include <sstream>

#include "DataTypes/Matrix44.h"

/**
 * Below you will find examples on how to handle requests from the robot control (e.g. user interface and program commands)
 * Also check main.cpp for examples on how to send requests from the app to the robot control.
 */

/**
 * @brief Constructor
 * @param target connection target in the following format: "hostname:port" or "ip:port", e.g. "localhost:5000"
 */
MinimalApp::MinimalApp(const std::string& target) : AppClient(std::string(APP_NAME), target) {}

/**
 * @brief Gets called on remote app function calls received from the robot control
 * @param function app function data
 */
void MinimalApp::AppFunctionHandler(const robotcontrolapp::AppFunction& function)
{
    // Select the app function to call
    if (function.name() == "pow")
    {
        ExampleExponentiation(function);
    }
    else
    {
        // Example on how to handle app functions (these are called by the program command "App")
        ExamplePrintAppFunctionParameters(function);
    }

    // Confirm that the function finished, otherwise the robot program will wait forever.
    // You may send this later if the function call takes some time but it must be sent at some point.
    SendFunctionDone(function.call_id());
	// Or call this in case the function failed. This stops the robot program.
	// SendFunctionFailed(function.call_id(), "failure reason");
}

/**
 * @brief Gets called on remote UI update requests received from the robot control
 * @param updates updated UI elements. Key is the element name, value contains the changes.
 */
void MinimalApp::UiUpdateHandler(const std::map<std::string, const robotcontrolapp::AppUIElement*>& updates)
{
    // Example on how to handle UI events (button clicked, value changed etc)
    ExamplePrintUIEvents(updates);

    // Example on how to handle UI events (button clicked, value changed etc)
    ExampleUIElementClicked(updates);
}

/**
 * @brief This example prints all app functions contained in an AppFunction request (e.g. coming from the App program command)
 * @param function
 */
void MinimalApp::ExamplePrintAppFunctionParameters(const robotcontrolapp::AppFunction& function) const
{
    std::cout << "App function '" << function.name() << "' called with call ID " << function.call_id() << ", label = '" << function.label() << "', ui hint = '"
              << function.ui_hint() << "', number of parameters = " << function.parameters_size() << std::endl;

    // print all parameters
    for (int i = 0; i < function.parameters_size(); i++)
    {
        const robotcontrolapp::AppFunction_Parameter& parameter = function.parameters().Get(i);
        std::ostringstream value;
        std::string typeStr;
        switch (parameter.value_case())
        {
            case robotcontrolapp::AppFunction_Parameter::kBoolValue:
                // boolean
                typeStr = "bool";
                value << (parameter.bool_value() ? "true" : "false");
                break;
            case robotcontrolapp::AppFunction_Parameter::kInt64Value:
                // integer
                typeStr = "int";
                value << parameter.int64_value();
                break;
            case robotcontrolapp::AppFunction_Parameter::kDoubleValue:
                // double
                typeStr = "double";
                value << parameter.double_value();
                break;
            case robotcontrolapp::AppFunction_Parameter::kStringValue:
                // string
                typeStr = "string";
                value << parameter.string_value();
                break;
            case robotcontrolapp::AppFunction_Parameter::kVector3Value:
                // 3 dimensional vector
                typeStr = "vector";
                value << "(" << parameter.vector3_value().x() << ", " << parameter.vector3_value().y() << ", " << parameter.vector3_value().z() << ")";
                break;
            case robotcontrolapp::AppFunction_Parameter::kCartesianValue:
                // cartesian position and orientation
                {
                    typeStr = "cartesian";
                    // read the values into a matrix, this allows easy access to the cartesian position and orientation
                    App::DataTypes::Matrix44 matrix(parameter.cartesian_value());
                    value << "X=" << matrix.GetX() << ", Y=" << matrix.GetY() << ", Z=" << matrix.GetZ() << ", A=" << matrix.GetA() << ", B=" << matrix.GetB()
                          << ", C=" << matrix.GetC();
                }
                break;
            default:
                break;
        }

        std::cout << "\tparameter '" << parameter.name() << "', type '" << typeStr << "', value '" << value.str() << "'" << std::endl;
    }
}

/**
 * @brief This example exponentiates the value of a number variable by a given exponent and assigns the result to a different number variable
 * @param function function parameters
 */
void MinimalApp::ExampleExponentiation(const robotcontrolapp::AppFunction& function)
{
    std::string baseVariableName;
    std::string resultVariableName;
    double exponentValue = 1;

    bool hasBaseVariable = false, hasResultVariable = false, hasExponent = false;

    // Get the function parameters by iterating over the list
    for (int i = 0; i < function.parameters_size(); i++)
    {
        const robotcontrolapp::AppFunction_Parameter& parameter = function.parameters().Get(i);

        // check parameter name and type
        if (parameter.name() == "base_variable" && parameter.value_case() == robotcontrolapp::AppFunction_Parameter::kStringValue)
        {
            baseVariableName = parameter.string_value();
            hasBaseVariable = true;
        }
        else if (parameter.name() == "result_variable" && parameter.value_case() == robotcontrolapp::AppFunction_Parameter::kStringValue)
        {
            resultVariableName = parameter.string_value();
            hasResultVariable = true;
        }
        else if (parameter.name() == "exponent_number" && parameter.value_case() == robotcontrolapp::AppFunction_Parameter::kDoubleValue)
        {
            exponentValue = parameter.double_value();
            hasExponent = true;
        }
    }

    // Check whether all parameters were received
    if (!hasBaseVariable || !hasResultVariable || !hasExponent)
    {
        std::cerr << "Function call \"exponentiation\" failed: incomplete function parameters!" << std::endl;
        return;
    }

    try
    {
        // Now request the value of the base number variable from the robot control
        std::shared_ptr<App::DataTypes::NumberVariable> baseVariable = GetNumberVariable(baseVariableName);
        double baseValue = baseVariable->GetValue();

        // And calculate the result
        double resultValue = pow(baseValue, exponentValue);

        // Write the result to the target variable
        SetNumber(resultVariableName, resultValue);

        // Do some debug output if you like
        std::cout << "Calculated " << baseValue << "^" << exponentValue << " = " << resultValue << ", result was written to variable \"" << resultVariableName
                  << "\"" << std::endl;
    }
    catch (std::exception& ex)
    {
        // Make sure to catch exceptions, e.g. when the variable does not exist!
        std::cerr << "Function call \"exponentiation\" failed: " << ex.what() << std::endl;
    }
}

/**
 * @brief This example prints all UI events contained in a UI update
 * @param updates UI updates
 */
void MinimalApp::ExamplePrintUIEvents(const std::map<std::string, const robotcontrolapp::AppUIElement*>& updates) const
{
    for (auto& update : updates)
    {
        switch (update.second->state().state_case())
        {
            case robotcontrolapp::AppUIElement::AppUIState::kButtonState:
                // button clicked
                {
                    bool isClicked = update.second->state().button_state() == robotcontrolapp::ButtonState::CLICKED;
                    std::cout << "Button '" << update.first << "' changed: is clicked = " << isClicked << std::endl;
                }
                break;
            case robotcontrolapp::AppUIElement::AppUIState::kCheckboxState:
                // checkbox toggled
                {
                    bool isChecked = update.second->state().checkbox_state() == robotcontrolapp::CheckboxState::CHECKED;
                    std::cout << "Checkbox '" << update.first << "' changed: is checked = " << isChecked << std::endl;
                }
                break;
            case robotcontrolapp::AppUIElement::AppUIState::kDropdownState:
                // drop down selected
                {
                    const std::string& selectedOption = update.second->state().dropdown_state().selected_option();
                    std::cout << "Dropdown '" << update.first << "' changed: selected option = " << selectedOption << std::endl;
                }
                break;
            case robotcontrolapp::AppUIElement::AppUIState::kImageState:
                // image clicked
                if (update.second->state().image_state().has_clicked_at())
                {
                    double x = update.second->state().image_state().clicked_at().x();
                    double y = update.second->state().image_state().clicked_at().y();
                    bool isClicked = update.second->state().image_state().clicked_at().is_clicked();
                    if (isClicked)
                    {
                        std::cout << "Image clicked '" << update.first << "': x = " << x << ", y = " << y << std::endl;
                    }
                }
                break;
            case robotcontrolapp::AppUIElement::AppUIState::kNumberfieldState:
                // number box changed
                {
                    double value = update.second->state().numberfield_state().current_number();
                    std::cout << "Number box '" << update.first << "' changed: value = " << value << std::endl;
                }
                break;
            case robotcontrolapp::AppUIElement::AppUIState::kTextfieldState:
                // text box changed
                {
                    const std::string& text = update.second->state().textfield_state().current_text();
                    std::cout << "Text box '" << update.first << "' changed: text = " << text << std::endl;
                }
                break;
            default:
                std::cerr << "Got UI update for UI element '" << update.first << "' with unknown type" << std::endl;
                break;
        }
    }
}

/**
 * @brief This example handles the plus/minus buttons to increase a text element in the app UI
 * @param updates UI updates
 */
void MinimalApp::ExampleUIElementClicked(const std::map<std::string, const robotcontrolapp::AppUIElement*>& updates)
{
    for (auto& update : updates)
    {
        // The following sample will increase/decrease a value if the plus/minus button is pressed and send it back to the UI.
        if (update.second->state().state_case() == robotcontrolapp::AppUIElement::AppUIState::kButtonState)
        {
            if (update.first == "buttonMinus") // button minus pressed
            {
                m_ExamplePlusMinusValue--;
                std::cout << "Decreasing UI number to " << m_ExamplePlusMinusValue << std::endl;
                SetNumber("textLeftRight", m_ExamplePlusMinusValue);
            }
            else if (update.first == "buttonPlus") // button plus pressed
            {
                m_ExamplePlusMinusValue++;
                std::cout << "Increasing UI number to " << m_ExamplePlusMinusValue << std::endl;
                SetNumber("textLeftRight", m_ExamplePlusMinusValue);
            }
        }
    }
}
